### PR TITLE
Enable select QU8 ARM [I]GEMM kernels with FP32 requantization

### PR DESCRIPTION
**Pull request by digantdesai**
Hello,

I am planning to explore using XNNPACK as a quantized backend for PyTorch on ARM CPUs. To get started with that, here with this PR, I am hoping to enable handful of QU8 ARM kernels with FP32 requantization scheme.
 
None of these kernels are enabled in PROD (may be except AArch64_ASM sources which do get added to PROD sources). Also, no changes in the src/init.c:init() has been introduced here, but would love to get some feedback from you on what’s a good way to initialize the library with a different requantization scheme for a given set of kernels e.g. QU8 w/ FP32 requantization.
 
I have also added appropriate tests for each of the kernels and tested the ARMv7 and ARMv8 Android CMake builds on a Pixel5. Below is a summary of that,

```
$ for arch in armeabi-v7a arm64-v8a; do for k in gemm igemm; do echo $arch - $k; adb shell "/data/local/tmp/${arch}/qu8-${k}-minmax-fp32-test --gtest_filter=-*SCALAR*" | sed -r "s/\..*//g" | sort | uniq -c | grep -e "OK\|PASSED\|SKIPPED\|FAILED"; done; done

armeabi-v7a - gemm
  30 [       OK ] QU8_GEMM_MINMAX_FP32_1X16C4__NEONDOT
  30 [       OK ] QU8_GEMM_MINMAX_FP32_1X16__NEONV8_MLAL_LANE
  30 [       OK ] QU8_GEMM_MINMAX_FP32_1X16__NEON_MLAL_LANE
  30 [       OK ] QU8_GEMM_MINMAX_FP32_1X8__NEON_MLAL_LANE
  30 [       OK ] QU8_GEMM_MINMAX_FP32_2X16C4__NEONDOT
  30 [       OK ] QU8_GEMM_MINMAX_FP32_4X16C4__NEONDOT
  30 [       OK ] QU8_GEMM_MINMAX_FP32_4X16__NEONV8_MLAL_LANE
  30 [       OK ] QU8_GEMM_MINMAX_FP32_4X16__NEON_MLAL_LANE
  30 [       OK ] QU8_GEMM_MINMAX_FP32_4X8__NEON_MLAL_LANE
   1 [  PASSED  ] 270 tests
armeabi-v7a - igemm
  30 [       OK ] QU8_IGEMM_MINMAX_FP32_1X16C4__NEONDOT
  30 [       OK ] QU8_IGEMM_MINMAX_FP32_1X16__NEONV8_MLAL_LANE
  30 [       OK ] QU8_IGEMM_MINMAX_FP32_1X16__NEON_MLAL_LANE
  30 [       OK ] QU8_IGEMM_MINMAX_FP32_1X8__NEON_MLAL_LANE
  30 [       OK ] QU8_IGEMM_MINMAX_FP32_2X16C4__NEONDOT
  30 [       OK ] QU8_IGEMM_MINMAX_FP32_4X16C4__NEONDOT
  30 [       OK ] QU8_IGEMM_MINMAX_FP32_4X16__NEONV8_MLAL_LANE
  30 [       OK ] QU8_IGEMM_MINMAX_FP32_4X16__NEON_MLAL_LANE
  30 [       OK ] QU8_IGEMM_MINMAX_FP32_4X8__NEON_MLAL_LANE
   1 [  PASSED  ] 270 tests
arm64-v8a - gemm
  30 [       OK ] QU8_GEMM_MINMAX_FP32_1X16C4__NEONDOT
  30 [       OK ] QU8_GEMM_MINMAX_FP32_1X16__NEONV8_MLAL_LANE
  30 [       OK ] QU8_GEMM_MINMAX_FP32_1X16__NEON_MLAL_LANE
  30 [       OK ] QU8_GEMM_MINMAX_FP32_1X8__NEON_MLAL_LANE
  30 [       OK ] QU8_GEMM_MINMAX_FP32_2X16C4__NEONDOT
  30 [       OK ] QU8_GEMM_MINMAX_FP32_4X16C4__AARCH64_NEONDOT_CORTEX_A55
  30 [       OK ] QU8_GEMM_MINMAX_FP32_4X16C4__AARCH64_NEONDOT_LD128
  30 [       OK ] QU8_GEMM_MINMAX_FP32_4X16C4__NEONDOT
  30 [       OK ] QU8_GEMM_MINMAX_FP32_4X16__NEONV8_MLAL_LANE
  30 [       OK ] QU8_GEMM_MINMAX_FP32_4X16__NEON_MLAL_LANE
  30 [       OK ] QU8_GEMM_MINMAX_FP32_4X8__NEON_MLAL_LANE
   1 [  PASSED  ] 330 tests
arm64-v8a - igemm
  30 [       OK ] QU8_IGEMM_MINMAX_FP32_1X16C4__NEONDOT
  30 [       OK ] QU8_IGEMM_MINMAX_FP32_1X16__NEONV8_MLAL_LANE
  30 [       OK ] QU8_IGEMM_MINMAX_FP32_1X16__NEON_MLAL_LANE
  30 [       OK ] QU8_IGEMM_MINMAX_FP32_1X8__NEON_MLAL_LANE
  30 [       OK ] QU8_IGEMM_MINMAX_FP32_2X16C4__NEONDOT
  30 [       OK ] QU8_IGEMM_MINMAX_FP32_4X16C4__AARCH64_NEONDOT_CORTEX_A55
  30 [       OK ] QU8_IGEMM_MINMAX_FP32_4X16C4__AARCH64_NEONDOT_LD128
  30 [       OK ] QU8_IGEMM_MINMAX_FP32_4X16C4__NEONDOT
  30 [       OK ] QU8_IGEMM_MINMAX_FP32_4X16__NEONV8_MLAL_LANE
  30 [       OK ] QU8_IGEMM_MINMAX_FP32_4X16__NEON_MLAL_LANE
  30 [       OK ] QU8_IGEMM_MINMAX_FP32_4X8__NEON_MLAL_LANE
   1 [  PASSED  ] 330 tests
```

I would love to get these merged and appreciate any feedback if you wish me to make any changes.

Thanks!

### CODE REVIEW

**Comment by Maratyszcza** [comment]
Why do you want to use FP32 requantization? ARM microkernels use RNDNU because it is faster.

**Comment by kimishpatel** [comment]
> Why do you want to use FP32 requantization? ARM microkernels use RNDNU because it is faster.

@Maratyszcza pytorch relies fp32 requantization scheme across the quantized backends.

By the way my understanding was fp32 requant was faster for aarch64. Is that not true?

**Comment by Maratyszcza** [comment]
FP32 requantization is faster on all platforms except ARM (both 32- and 64-bit), where RNDNU is faster. You can try it yourself with `qu8-requantization-bench` from XNNPACK.

**Comment by digantdesai** [comment]
@Maratyszcza - thanks for a quick feedback. Also thanks @kimishpatel for chiming in.

Independent of the performance of these methods, as Kimish said, the requantization scale value restrictions arising from gemmlowp/rndnu assumptions for doing fixed point scale multiplication _might be_ too restrictive for PyTorch models expecting fp32 requantization _everywhere_. To me int-only requantization is appealing and would like to leverage that in the future. For now, it might be a easier for us to get started with XNNPACK with fp32 requantization schema. @kimishpatel please feel free to chime in on this.

Thoughts?



Also as you suggested I ran `qu8-requantization-bench` on a Pixel-5, and to my surprize I consistently get lower performance with rndnu. This is by no means a detailed benchmarking, just a quick and dirty numbers. The arm64 binary was build using CMake and ndk-21e.

```
$ taskset 80 ./qu8-requantization-bench --benchmark_repetitions=5 --benchmark_filter=.*neon.*
2021-12-02T09:16:10-08:00
Running ./qu8-requantization-bench
Run on (8 X 1804.8 MHz CPU s)
***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
-----------------------------------------------------------------------------------------------
Benchmark                                     Time             CPU   Iterations UserCounters...
-----------------------------------------------------------------------------------------------
Requantization/fp32__neon                  6531 ns         6518 ns       103470 bytes_per_second=9.21262G/s cpufreq=2.4G items_per_second=1.9784G/s
Requantization/fp32__neon                  6531 ns         6518 ns       103470 bytes_per_second=9.21317G/s cpufreq=2.4G items_per_second=1.97851G/s
Requantization/fp32__neon                  6532 ns         6519 ns       103470 bytes_per_second=9.21244G/s cpufreq=2.4G items_per_second=1.97836G/s
Requantization/fp32__neon                  6531 ns         6518 ns       103470 bytes_per_second=9.21358G/s cpufreq=2.4G items_per_second=1.9786G/s
Requantization/fp32__neon                  6531 ns         6518 ns       103470 bytes_per_second=9.21342G/s cpufreq=2.4G items_per_second=1.97857G/s
Requantization/fp32__neon_mean             6531 ns         6518 ns            5 bytes_per_second=9.21305G/s cpufreq=2.4G items_per_second=1.97849G/s
Requantization/fp32__neon_median           6531 ns         6518 ns            5 bytes_per_second=9.21317G/s cpufreq=2.4G items_per_second=1.97851G/s
Requantization/fp32__neon_stddev          0.399 ns        0.352 ns            5 bytes_per_second=522.04k/s cpufreq=0 items_per_second=106.914k/s
Requantization/gemmlowp__neon              6579 ns         6567 ns       106604 bytes_per_second=9.14483G/s cpufreq=2.4G items_per_second=1.96384G/s
Requantization/gemmlowp__neon              6578 ns         6567 ns       106604 bytes_per_second=9.14507G/s cpufreq=2.4G items_per_second=1.96389G/s
Requantization/gemmlowp__neon              6578 ns         6567 ns       106604 bytes_per_second=9.14495G/s cpufreq=2.4G items_per_second=1.96386G/s
Requantization/gemmlowp__neon              6579 ns         6567 ns       106604 bytes_per_second=9.14414G/s cpufreq=2.4G items_per_second=1.96369G/s
Requantization/gemmlowp__neon              6578 ns         6567 ns       106604 bytes_per_second=9.14503G/s cpufreq=2.4G items_per_second=1.96388G/s
Requantization/gemmlowp__neon_mean         6578 ns         6567 ns            5 bytes_per_second=9.14481G/s cpufreq=2.4G items_per_second=1.96383G/s
Requantization/gemmlowp__neon_median       6578 ns         6567 ns            5 bytes_per_second=9.14495G/s cpufreq=2.4G items_per_second=1.96386G/s
Requantization/gemmlowp__neon_stddev      0.674 ns        0.276 ns            5 bytes_per_second=402.661k/s cpufreq=0 items_per_second=82.4651k/s
Requantization/rndna__neon                10697 ns        10677 ns        65560 bytes_per_second=5.62415G/s cpufreq=2.4G items_per_second=1.20778G/s
Requantization/rndna__neon                10696 ns        10677 ns        65560 bytes_per_second=5.62415G/s cpufreq=2.4G items_per_second=1.20778G/s
Requantization/rndna__neon                10697 ns        10678 ns        65560 bytes_per_second=5.62411G/s cpufreq=2.4G items_per_second=1.20777G/s
Requantization/rndna__neon                10697 ns        10678 ns        65560 bytes_per_second=5.62409G/s cpufreq=2.4G items_per_second=1.20776G/s
Requantization/rndna__neon                10697 ns        10678 ns        65560 bytes_per_second=5.62409G/s cpufreq=2.4G items_per_second=1.20776G/s
Requantization/rndna__neon_mean           10697 ns        10678 ns            5 bytes_per_second=5.62412G/s cpufreq=2.4G items_per_second=1.20777G/s
Requantization/rndna__neon_median         10697 ns        10678 ns            5 bytes_per_second=5.62411G/s cpufreq=2.4G items_per_second=1.20777G/s
Requantization/rndna__neon_stddev         0.445 ns        0.052 ns            5 bytes_per_second=28.8543k/s cpufreq=0 items_per_second=5.9094k/s
```

**Comment by Maratyszcza** [comment]
Ok, I think we can get it merged and continue discussion

**Review by Maratyszcza — APPROVED** [review]

**Comment by digantdesai** [comment]
Thanks @Maratyszcza!

Please let me know if you have any thought/preference on updating init() to allow initializing the library with a different requantization scheme.

**Comment by Maratyszcza** [comment]
We've relaxed restrictions on requantization scales earlier this week, now scales up to 256 are supported. Does it satisfy your use-case?

If you still would like this patch in, please rebase and re-generate microkernels.

**Comment by kimishpatel** [comment]
> We've relaxed restrictions on requantization scales earlier this week, now scales up to 256 are supported. Does it satisfy your use-case?
> 
> If you still would like this patch in, please rebase and re-generate microkernels.

Hi Marat,
This is definitely great. But we might have to 1) test and make sure this works for our use cases and 2) internally align that going back to integer arithmetic based requantization instead of fp32 is ok. I am more concerned on the 2, especially because the two backends pytorch supports currently are aligned to fp32 requantization at the moment.

So if it is ok with you, we can merge this PR and then revisit. I would definitely love to make use of perf advantages.

Also, I did see your PRs relaxing the requirement. If you have any pointers on why moving from vshlq_s32 to vqshlq_s32 supports new restriction on scale, it would be great.

File: BUILD.bazel [hunk 1]
```diff
@@ -3932,15 +3939,18 @@ ALL_NEONDOT_MICROKERNEL_SRCS = [
...
     "src/qu8-igemm/gen/1x32c4-minmax-rndnu-neondot.c",
     "src/qu8-igemm/gen/2x8c4-minmax-rndnu-neondot.c",
+    "src/qu8-igemm/gen/2x16c4-minmax-fp32-neondot.c",
     "src/qu8-igemm/gen/2x16c4-minmax-rndnu-neondot.c",
     "src/qu8-igemm/gen/2x32c4-minmax-rndnu-neondot.c",
     "src/qu8-igemm/gen/3x8c4-minmax-rndnu-neondot.c",
     "src/qu8-igemm/gen/3x16c4-minmax-rndnu-neondot.c",
     "src/qu8-igemm/gen/3x32c4-minmax-rndnu-neondot.c",
     "src/qu8-igemm/gen/4x8c4-minmax-rndnu-neondot.c",
+    "src/qu8-igemm/gen/4x8c4-minmax-rndnu-neondot.c",
```
Comment (Maratyszcza) [inline BUILD.bazel:3953]: Duplicate. Is it supposed to be "src/qu8-igemm/gen/4x8c4-minmax-fp32-neondot.c"?

Comment (digantdesai) [reply BUILD.bazel:3953]: Good catch, yeah it's the latter. I never tested with Bazel, let me fix it, thanks!

**Comment by Maratyszcza** [comment]
Fails to build due to missing input file `src/qu8-igemm/gen/4x8c4-minmax-fp32-neondot.c`

File: BUILD.bazel [hunk 2]
```diff
@@ -3934,14 +3941,17 @@ ALL_NEONDOT_MICROKERNEL_SRCS = [
...
     "src/qu8-igemm/gen/1x16c4-minmax-rndnu-neondot.c",
     "src/qu8-igemm/gen/1x32c4-minmax-rndnu-neondot.c",
     "src/qu8-igemm/gen/2x8c4-minmax-rndnu-neondot.c",
+    "src/qu8-igemm/gen/2x16c4-minmax-fp32-neondot.c",
     "src/qu8-igemm/gen/2x16c4-minmax-rndnu-neondot.c",
     "src/qu8-igemm/gen/2x32c4-minmax-rndnu-neondot.c",
     "src/qu8-igemm/gen/3x8c4-minmax-rndnu-neondot.c",
     "src/qu8-igemm/gen/3x16c4-minmax-rndnu-neondot.c",
     "src/qu8-igemm/gen/3x32c4-minmax-rndnu-neondot.c",
+    "src/qu8-igemm/gen/4x8c4-minmax-fp32-neondot.c",
```
Comment (Maratyszcza) [inline BUILD.bazel:3954]: Note: this doesn't match the GEMM microkernel (which is `4x16c4`)

Comment (digantdesai) [reply BUILD.bazel]: ah, yet another Basel typo :\

**Pull request merged**